src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -530,12 +533,21 @@ DistributedModifyShardInterval(Query *query)
 	restrictClauseList = QueryRestrictList(query);
 	prunedShardList = PruneShardList(distributedTableId, tableId, restrictClauseList,
 									 shardIntervalList);
-
-	if (list_length(prunedShardList) != 1)
+	prunedShardCount = list_length(prunedShardList);
+	if (prunedShardCount != 1)
 	{
-		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("distributed modifications must target exactly one "
-							   "shard")));
+		if (selectTask)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
```

Comment:
To clarify, this condition only gets triggered if a hash-partitioned table is improperly distributed? I actually ran into it in the hash_pruning regression tests while rebasing explain.